ccp/index_reader.py

```diff
@@ -449,33 +502,116 @@ def run(self):
         )
 
         if stale_projects:
-            print ("List of stale projects:\n{}".format(
+            _print("List of stale projects:\n{}".format(
                 "\n".join(stale_projects)))
             # delete all the stal projects/buildconfigs
             self.bc_manager.delete_buildconfigs(stale_projects)
 
-        # oc process and oc apply to all fresh and existing jobs
+        _print("Number of projects to be updated/created: {}".format(
+            len(index_projects)))
+
+        self.batch_process_projects(
+            index_projects,
+            batch_size,
+            polling_interval,
+            batch_outstanding_builds_cap
+        )
+
+    def batch(self, target, batch_size):
+        """
+        Returns a generator object yielding chunks of list
+        of length=batch_size from target list
+        """
+        for i in range(0, len(target), batch_size):
+            yield target[i:i + batch_size]
+
+    def batch_process_projects(self,
+                               index_projects, batch_size, polling_interval,
+                               batch_outstanding_builds_cap):
+        """
+        Given a list of projects, oc apply the buildconfigs
+        in batches of size $batch_size and oc apply the weekly
+        scan jobs at the end.
+        Watch the outstanding builds at $polling_interval and
+        schedule next batch if current outstanding jobs are
+        less than or equal to $batch_outstanding_builds_cap.
+        """
+        # Split the projects to process in equal sized chunks
+        generator_obj = self.batch(index_projects, batch_size)
+
+        _print("Starting index processing with\n"
+               "Batch size={}\nBatch polling Interval (in seconds)={}\n"
+               "Batch outstndaing builds cap count={}\n".format(
+                   batch_size,
+                   polling_interval,
+                   batch_outstanding_builds_cap))
+        # iterate over batches of projects to be processed
+        for batch in generator_obj:
+            # Check if builds are in status.phase other than Complete
+            # or Failed. We dont care about builds which are failed
+            # or complete to queue up next batch of jobs to process
+            outstanding_builds = self.bc_manager.list_builds_except(
+                status=["Complete", "Failed"],
+                filter_builds=self.infra_projects)
+
+            # wait until current outstanding builds are more than
+            # configured cap for outstanding builds
+            while len(outstanding_builds) > batch_outstanding_builds_cap:
+                _print("Waiting for completion of builds {}\n".format(
+                    outstanding_builds))
+                time.sleep(polling_interval)
+
+                outstanding_builds = self.bc_manager.list_builds_except(
+                    status=["Complete", "Failed"],
+                    filter_builds=self.infra_projects)
```

Comment:
IIUC, we're executing a shell command to get output here. `run_cmd` and `list_builds_except` don't seem to have exception handling. So, if fetching outstanding builds fails due to some issue (like failure in executing `oc` command), seed-job would fail. Correct me if I'm wrong.

If I'm correct, I think this part is deep into the execution of our service. We could give a few retries to the scenario where `oc` fails to get our desired task done and fail the seed-job appropriate message if it fails even after the retries.

Thoughts @navidshaikh @bamachrn @mohammedzee1000 ?

Reply:
I think it makes sense - We can keep a retry cycle of 30 seconds with timeout limit set to 5 minutes ? And fail it is 5 minutes (10 retries) are lapsed ?

Reply:
I am adding the retries in the actual method itself, as its being called multiple times.

Reply:
It makes sense. It applies to all the operations performed via `oc`. Adding a retry decorator which can be re-used and have it applied to al oc related operations.